### DISCUSSION TITLE
Refine for object when using zod not work without optional

### DISCUSSION THREAD

**Discussion by duynd243** [colinhacks/zod#2071] [Q&A] [answered]
My Codesandbox link: https://codesandbox.io/s/react-typescript-forked-uktywz

Can anyone help me, why the validation for "End date must be after start date" not work when I remove optional() on the something key of zod schema? The problem also occurs when something is an array.

**Comment by JacobWeisenburger** [comment, marked as answer]
Everything seems to be working as expected for me.
```ts
const schema = z.object( {
    something: z.object( { key: z.string() } ),
    name: z.string().min( 1, 'Name is required' ),
    startDate: z.coerce.date(),
    endDate: z.coerce.date()
} ).refine(
    data => data.startDate < data.endDate,
    {
        message: 'End date must be after start date',
        path: [ 'endDate' ],
    }
)

const result = schema.safeParse( {
    name: 'John',
    startDate: '2023-02-21',
    endDate: '2023-02-20',
    something: { key: 'value' }
} )
!result.success && console.log( result.error.issues )
// [
//     {
//         code: 'custom',
//         message: 'End date must be after start date',
//         path: [ 'endDate' ]
//     }
// ]
```

Can you explain specifically what you mean by 'not work'?

**Reply by duynd243** [reply]
Thank you for your response Jacob, I didn't test with safeParse method. Now I found what the problem is, It because that I didn't regrister all the required field from the zod schema with react-hook-form.

**Reply by JacobWeisenburger** [reply]
If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by manhhung-fpt** [reply]
I got same problem, thank you Dinh Dy!.
